Page cache counters for separate transaction

MishaDemianenko opened the pull request:
Previously page cache counters and metrics like number of flushes, bytes read etc was global and there was no possibility to find out in what way particular transaction or query influence it.

This PR will splits tracing into page cache global and thread local tracing of page cursors.
Also making thread local information available for consuming by any interesting parties to be able to report any transaction based page cache counters.
In opposed to previous behaviour when all events where reported directly to global tracer,  in new model  events will be reported first to thread local tracer and then eventually those will be propagated to global tracer.

MishaDemianenko committed:
Split page cache tracer into page cache tracer and page cursor tracer. Cleanup interfaces. Tests are in best case compilable.

MishaDemianenko committed:
Introduce way of injecting cursor tracers into page cursors from the outside.
Allow specify tracers type on platform creation during database start.

MishaDemianenko committed:
Move creation of page cursor tracer supplier to Tracers.
Allow access to current cursor tracer from kernel transaction statement.

MishaDemianenko committed:
Adapt and rebuild page cache test that use recording tracers infrastructure.
Make current tests pass with new combination of page cache and page cursor tracers.

MishaDemianenko committed:
Fix page cache rule config import

MishaDemianenko committed:
Update page cache stress builder to be cursor tracing aware.
Update javadoc of RecordingPageCursorTracer.
Use plain TestDirectory in MuninnPageCacheStressIT instead of temp directory.

MishaDemianenko committed:
Refactor linear page cache history tracer to have separate tracers for global and cursor local events.
Add support for cursor events tracers.

MishaDemianenko committed:
Fix new tests after rebase

MishaDemianenko committed:
Adapt test to new way of pin events tracking.

MishaDemianenko committed:
Introduce Page Cursor Counters interface to split
cursor counters and tracers to isolate code that can only access counters
and actually do cursor events tracing.

MishaDemianenko committed:
Expand number of exposed counters in page cursor counters.
Tests for page cursor reported counters

MishaDemianenko committed:
Remove switch to enable/disable pin events and trace them always.

MishaDemianenko committed:
Trace and count eviction exception that happen on page eviction
while operating with page specific page cursor

MishaDemianenko committed:
Update initialization of page cursor tracers, reset counters on init
tests for reset and incorrect reported unpin value.

MishaDemianenko committed:
Convert todo into card item

MishaDemianenko committed:
Add integration test for page cursor tracer and page cache tracer cooperation and counters propagation.

MishaDemianenko committed:
Use LongAdders in DefaultPageCacheTracer to accumulate total counter value,
update way of evaluation current cycle and total counter in page cursor tracer.

chrisvest commented:
@MishaDemianenko Needs a rebase.

MishaDemianenko commented:
@chrisvest done

tinwelint commented:
You could rebase this PR to hopefully get it green.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCursor.java:
```diff
@@ -72,16 +71,18 @@
     private Object cursorException;
 
-    MuninnPageCursor( long victimPage )
+    private final PageCursorTracer pageCursorTracer;
```
Put it next to the `victimPage` field.

chrisvest replied:
It can also still be called `tracer`, for a shorter name.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/tracing/PageCacheTracer.java:
```diff
@@ -170,4 +199,20 @@ public String toString()
+
+    void bytesWritten( long bytesWritten );
+
+    void flushes( long flushes );
```
Give them a bit of javadocs.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/tracing/cursor/DefaultPageCursorTracer.java:
```diff
@@ -0,0 +1,295 @@
+
+    private PageCacheTracer pageCacheTracer;
+
+    public void init( PageCacheTracer pageCacheTracer )
```
This needs an `@Override` doesn't it?

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/tracing/cursor/DefaultPageCursorTracer.java:
```diff
@@ -0,0 +1,295 @@
+        this.pageCacheTracer = pageCacheTracer;
+    }
+
+    public void reportEvents()
```
Some javadocs would be nice, to clarify what this does and doesn't do.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/tracing/cursor/PageCursorTracer.java:
```diff
@@ -0,0 +1,112 @@
+
+    void init( PageCacheTracer tracer );
+
+    void reportEvents();
```
I see, _this_ is where the javadoc is supposed to be.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCursor.java:
```diff
@@ -131,6 +132,7 @@ public final void close()
             if ( cursor.pagedFile != null )
             {
                 cursor.unpinCurrentPage();
+                cursor.pageCursorTracer.reportEvents();
```
I'm not entirely convinced that this is the right place to put it. Cypher wants to get numbers for a whole Statement, in which case it makes sense to collect and report the data when a statement is closed. Linked cursors would presumably also be attached to the same thread-local tracer, in which case there is no point to do the reporting several times.

MishaDemianenko replied:
There are couple of reasons why i did it like that:
- if statement is quite big and long running we will not gonna be able to see any page cache counters update and that will make tracing and understanding current situation non clear and hard to trace since all counters will be reported only once in the end when statement will be closed.
- currently code is really bad on closing statement (that is a some code path problem and should be fixed)

Agree on the point of linked cursors: if we will assume that our page cursors tracers are thread local and all of the cursors have the same tracer we should avoid doing reporting multiple times.
My thinking was a bit different - local counters in cursor remember what was done so far in  each separate tracer, so we can get statement dependent info now.

But agree your proposal will make code a bit simpler and faster even if two mentioned above points will not gonna be solved, but we can come back to them later if that will be a problem. Will do like you suggest.

chrisvest reviewed (changes requested):
Had a few comments, but in particular I think we need to take a closer look at where the cursor tracer reporting is done.

chrisvest committed:
Small page cursor tracer cleanups

MishaDemianenko committed:
Java docs

MishaDemianenko committed:
Report tracing events on statement force close,
forcefully close statement after transaction commit or rollback only to catch all tracing events that can happen during transaction close.

MishaDemianenko commented:
@chrisvest updated and another benchmark run triggered to see where we in terms of performance with this update.

chrisvest commented:
@MishaDemianenko Very nice.
